File: PaperLib/Models/PaperCategorizer.swift
//
//  PaperTag.swift
//  PaperLib
//
//  Created by GeoffreyChen on 26/11/2021.
//

import Foundation
import RealmSwift

class IDObject: Object {
    @objc dynamic var _id: ObjectId

    override init() {
        self._id = ObjectId.generate()
    }

    override class func primaryKey() -> String? {
        return "_id"
    }

    override class func shouldIncludeInDefaultSchema() -> Bool {
        self != IDObject.self
    }
}

class PaperCategorizer: IDObject {
    @Persisted var _partition: String?
    @Persisted var count: Int = 0
    @Persisted var name: String = ""

    convenience required init(name: String) {
        self.init()
        self.name = name
    }
}

class PaperTag: PaperCategorizer, ObjectKeyIdentifiable {
}

class PaperFolder: PaperCategorizer, ObjectKeyIdentifiable {
}

File: src/models/PaperCategorizer.ts
import { ObjectId } from 'bson';

export interface PaperCategorizer {
  _id: ObjectId;
  _partition: string;
  name: string;
  count: number;
}

export interface PaperTag extends PaperCategorizer {
  _id: ObjectId;
  _partition: string;
  name: string;
  count: number;
}

export interface PaperFolder extends PaperCategorizer {
  _id: ObjectId;
  _partition: string;
  name: string;
  count: number;
}

export type CategorizerType = 'PaperTag' | 'PaperFolder';

export const PaperTagSchema = {
  name: 'PaperTag',
  primaryKey: '_id',
  properties: {
    _id: 'objectId',
    _partition: 'string?',
    name: 'string',
    count: 'int',
  },
};

export const PaperFolderSchema = {
  name: 'PaperFolder',
  primaryKey: '_id',
  properties: {
    _id: 'objectId',
    _partition: 'string?',
    name: 'string',
    count: 'int',
  },
};
